### PR TITLE
ENH: `stats.chisquare`: add `sum_check` to disable check that total observed/expected counts are equal

### LINKED ISSUE #12282: API: stats.chisquare: must observed and expected frequencies add to same total?
The function scipy.stats.chisquare performs the chi-square test on a vector of observed and expected frequencies.  For the test to make sens, the observed and expected frequency vectors must sum to the same total, otherwise the result is nonsense since the inputs are incompatible.  So there are two options: either an error should be thrown if they do not, or f_exp need to be rescaled so they sum to the total of f_obs.  The first is probably better, or the second with a warning thrown.

#### Reproducing code example:

```
# In the following example, one vector is an exact multiple of the other.
# this means that the observed and expected frequencies are exactly
# proportional.  This should give a p-value of 1 (not significant at all).  
# Instead, you get the following:
from scipy.stats import chisquare
chisquare(f_obs=[10,20], f_exp=[30,60])

# Power_divergenceResult(statistic=40.0, pvalue=2.5396285894708634e-10)
# the statistic of 40 is calculated as following, which directly follows the 
# formula.  This is then plugged into a chi squared distribution to get a 
#p-value of close to 0, which is the opposite of the significance you should
# get.
# ((10-30)**2 /30) + ((20-60)** 2 / 60) = 40

# Instead, here is what should happen
import numpy as np
fobs = np.array([10,20])
fexp = np.array([30,60])
# adjust the totals
# gives array([10., 20.]), the same as observed
fexp = fexp * (np.sum(fobs)/np.sum(fexp)) 
chisquare(f_obs=fobs, f_exp=fexp)
# the correct result
# Power_divergenceResult(statistic=0.0, pvalue=1.0)
```

#### Scipy/Numpy/Python version information:
```
1.4.1 1.18.4 sys.version_info(major=3, minor=7, micro=4, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-12282

#### What does this implement/fix?
gh-12282 noted that the sums of observed and expected counts must be equal for `scipy.stats.chisquare` to produce accurate p-values for Pearson's chi-squared test.

gh-13193 added a check that these sums are equal, raising an exception otherwise. 

The change was quite disruptive (e.g. see gh-14298), but also seems to have caught real user errors (e.g. https://github.com/scipy/scipy/issues/14298#issuecomment-1031901653).

Recently, https://github.com/scipy/scipy/issues/12282#issuecomment-2278421287 noted a legitimate use case for the function even when the total counts are not equal. 

This PR adds an argument to disable the sum check and adjusts the documentation to explain the rationale for the check itself and for opting out of the check.

#### Additional information 
~~Need to add tests.~~ Done.

### CODE REVIEW

File: scipy/stats/_stats_py.py [hunk 1]
```diff
@@ -7166,29 +7172,35 @@ def warn_masked(arg):
...
-def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
-    """Calculate a one-way chi-square test.
+def chisquare(f_obs, f_exp=None, ddof=0, axis=0, *, sum_check=True):
+    """Perform Pearson's chi-squared test.
 
-    The chi-square test tests the null hypothesis that the categorical data
-    has the given frequencies.
+    Pearson's chi-squared test [1]_ assesses the null hypothesis that the observed
+    frequencies (counts) are a sample from a multinomial distribution; that is,
+    they are the counts after *N* total observations from a categorical
```
Comment (josef-pkt) [inline scipy/stats/_stats_py.py:7180]: "after" here sounds strange to me
either
"after N total observations are taken from ..."
or "after" -> "conditional on"

**Comment by josef-pkt** [comment]
LGTM

**Comment by mdhaber** [comment]
Great. I changed it to "after N total observations are taken from..." as requested.

I also added a unit test and attempted to compactly motivate the chi-squared test on Poisson data, since I couldn't easily find a reference.

**Comment by mdhaber** [comment]
@HDembinski @petrelharp does this look good to allow for use of `stats.chisquare` beyond Pearson's chi-squared test?

File: scipy/stats/_stats_py.py [hunk 2]
```diff
@@ -7231,13 +7239,23 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
     the asymptotic distribution is not chi-square, in which case this test
     is not appropriate.
```
Comment (HDembinski) [inline scipy/stats/_stats_py.py:7240]: Suggestion: However, it is also possible that the asymptotic distribution for an arbitrary estimation is not chi-square (for example, because the estimation is not consistent), in which case this test is not appropriate.

Comment (HDembinski) [reply scipy/stats/_stats_py.py:7240]: I think it should be made clear that the statistic is asymptotically always chi-square distributed if one indeed passes observed and expected frequencies, and no estimation is happening. Deviations from an asymptotic chi-square distribution can occur if the estimation is not consistent.

**Review by HDembinski — APPROVED** [review]
For me it looks fine. I would prefer a bit more clarification when the test statistic is not asymptotically chi-square distributed (see inline comment).

File: scipy/stats/_stats_py.py (hunk 2 above)
Comment (mdhaber) [reply scipy/stats/_stats_py.py:7240]: Thanks. Since that's part of the existing docstring and not altered here, I'll leave that for another pull request.

File: scipy/stats/_stats_py.py [hunk 3]
```diff
@@ -7166,29 +7172,35 @@ def warn_masked(arg):
...
-    """Calculate a one-way chi-square test.
+def chisquare(f_obs, f_exp=None, ddof=0, axis=0, *, sum_check=True):
+    """Perform Pearson's chi-squared test.
 
-    The chi-square test tests the null hypothesis that the categorical data
-    has the given frequencies.
+    Pearson's chi-squared test [1]_ assesses the null hypothesis that the observed
+    frequencies (counts) are a sample from a multinomial distribution; that is,
+    they are the counts after *N* total observations are taken from a categorical
+    distribution.
```
Comment (petrelharp) [inline scipy/stats/_stats_py.py:7181]:
```suggestion
    Pearson's chi-squared test [1]_ is a goodness-of-fit test for a multinomial
    distribution with given probabilities; that is, it assesses the null hypothesis
    that the observed  frequencies (counts) are obtained by independent
    sampling of *N* observations with the given expected frequencies.
```

I'm not sure if that's better - in particular, I'm trying to make more clear that the test isn't just asking whether they follow *any* multinomial, but rather the particular multinomial with `f_exp` probabilities.

File: scipy/stats/_stats_py.py [hunk 4]
```diff
@@ -7214,15 +7226,11 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
     -----
     This test is invalid when the observed or expected frequencies in each
     category are too small.  A typical rule is that all of the observed
```
Comment (petrelharp) [inline scipy/stats/_stats_py.py:7228]:
```suggestion
    This test is not well-calibrated when the observed or expected frequencies in each
    category are too small.  A typical rule is that all of the observed
```
Perhaps more precise here? "Invalid" doesn't have a statistical meaning.

File: scipy/stats/_stats_py.py [hunk 5]
```diff
@@ -7214,15 +7226,11 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
...
-    and expected frequencies should be at least 5. According to [3]_, the
-    total number of samples is recommended to be greater than 13,
+    and expected frequencies should be at least 5. According to [2]_, the
+    total number of observations is recommended to be greater than 13,
     otherwise exact tests (such as Barnard's Exact test) should be used
     because they do not overreject.
```
Comment (petrelharp) [inline scipy/stats/_stats_py.py:7232]:
```suggestion
    otherwise exact tests (such as Barnard's Exact test or a permutation test) should be used
    because they do not overreject.
```
Might as well point them towards the more general right way to do it (Barnard's test is for 2x2 tables). Note that you can't just permute the counts, but rather the categorical data the counts came from; this should point them in the right direction.

File: scipy/stats/_stats_py.py (hunk 2 above)
Comment (petrelharp) [reply scipy/stats/_stats_py.py:7240]: I'd advocate for not getting into the weeds here - we're not trying to provide a textbook on statistical theory?  The caveat that "this test may not be appropriate" applies to any procedure at all. If you've got a particular gotcha that's easy to explain in mind, though, that could be useful? (Note that in my edit above I mentioned that it's testing whether the observations are independent, which is the main source of discrepancy; we could explain more but maybe anyone who's sophisticated enough to pick up on that will already be familiar with the chi-squared test?)

File: scipy/stats/_stats_py.py (hunk 4 above)
Comment (josef-pkt) [reply scipy/stats/_stats_py.py:7228]: I see "invalid" quite often in the statistics literature.

The main violation is that the size (rejection frequency under the null) is not maintained, if assumption do not hold.

It's less precise than other terms, but we don't need to specify what exactly will go wrong under different violations of assumptions.

File: scipy/stats/_stats_py.py [hunk 6]
```diff
@@ -7231,13 +7239,23 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
     the asymptotic distribution is not chi-square, in which case this test
     is not appropriate.
 
+    For Pearson's chi-squared test, the total observed and expected counts must match
+    for the p-value to accurately reflect the probability of observing such an extreme
+    value of the statistic under the null hypothesis.
```
Comment (petrelharp) [inline scipy/stats/_stats_py.py:7244]:
```suggestion
    Pearson's chi-squared test assumes the total number of observations is fixed (thus
    giving the default ``k-1`` degrees of freedom), and so an error will be generated if
    the total observed and expected counts are not equal; set ``sum_check=False`` to
    disable this check.
```

File: scipy/stats/_stats_py.py [hunk 7]
```diff
@@ -7231,13 +7239,23 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
...
+    This function may be used to perform other statistical tests that do not require
+    the total counts to be equal. For instance, to test the null hypothesis that
+    ``f_obs[i]`` is Poisson-distributed with expectation ``f_exp[i]``, set ``ddof=-1``
+    and ``sum_check=False``. This test follows from the fact that a Poisson random
+    variable with mean and variance ``f_exp[i]`` is approximately normal with the
+    same mean and variance; the chi-squared statistic standardizes, squares, and sums
+    the observations; and the sum of ``n`` squared standard normal variables follows
+    the chi-squared distribution with ``n`` degrees of freedom.
```
Comment (petrelharp) [inline scipy/stats/_stats_py.py:7252]: Up to you, but I don't think we actually need the derivation here? (I'm sorry I don't have a reference for the application, but it's one of those things where a good reference would be "any statistician knows this".)

File: scipy/stats/_stats_py.py (hunk 6 above)
Comment (mdhaber) [reply scipy/stats/_stats_py.py:7244]: Could you summarize the intent of the change? The intent of the original is just to set the stage for the rest of the paragraph, not document `sum_check` or `ddof`.

File: scipy/stats/_stats_py.py (hunk 5 above)
Comment (mdhaber) [reply scipy/stats/_stats_py.py:7232]: Sure, you're welcome to open a PR with that information. This PR does not intend to adjust much more than necessary to document the new parameter and why it might be used.

File: scipy/stats/_stats_py.py (hunk 3 above)
Comment (mdhaber) [reply scipy/stats/_stats_py.py:7181]: I think it is good to start with describing it as a goodness of fit test rather than jumping straight into the null hypothesis, and yes, good that we're interested in a particular multinomial distribution.

The one thing i'd take out is the parenthetical bit about the default, and I'd add back the part about the `N` observations being from *a categorical distribution*. (I'd say the input is *one* observation from a multinomial distribution parametrized by total count `N` and the provided frequencies.)

File: scipy/stats/_stats_py.py (hunk 7 above)
Comment (mdhaber) [reply scipy/stats/_stats_py.py:7252]: Well, yeah, I included the sentence only because we didn't have a reference.

File: scipy/stats/_stats_py.py (hunk 4 above)
Comment (petrelharp) [reply scipy/stats/_stats_py.py:7228]: You're right, "invalid" is used; apologies - I meant that it doesn't have a precise meaning.  And, it implies that tests are either "valid" or "not valid"; when in practice most tests (including this one) are approximate - it's a case of "all models are false but some are useful". How about 
```
This test is not accurate ...
```
... ?

One reason I'm suggesting this is because it's very common for people to ask things like "is it allowed to do this test"; as if that's a yes-no question, rather than asking "how useful will this be".

File: scipy/stats/_stats_py.py (hunk 3 above)
Comment (petrelharp) [reply scipy/stats/_stats_py.py:7181]: Edited - how's that?

File: scipy/stats/_stats_py.py (hunk 6 above)
Comment (petrelharp) [reply scipy/stats/_stats_py.py:7244]: Well, the intent was to say why they 'must match' (the reason being the degrees of freedom), and to more explicitly give the reasoning behind the sum_check. However, what you had there was good, so if it is more opaque to you, then ignore my suggestion.

File: scipy/stats/_stats_py.py (hunk 7 above)
Comment (petrelharp) [reply scipy/stats/_stats_py.py:7252]: Alternatively, you could just say more concisely that

> This test follows from asymptotic Normality of the Poisson distribution.

That's not a reference, but gives keywords that would end people up in the right place?

(As for the t-test: I think the point of a reference in this sort of thing is to provide further documentation on (a) what exactly is being done, and (b) why; I don't think it's important to, say, reference Gosset's original paper.)

Comment (mdhaber) [reply scipy/stats/_stats_py.py:7252]: > As for the t-test:

That is correct.

File: scipy/stats/_stats_py.py (hunk 6 above)
Comment (mdhaber) [reply scipy/stats/_stats_py.py:7244]: The suggestion is good content; it's just different information from what I had in in mind for the beginning of this paragraph. Thanks!

File: scipy/stats/_stats_py.py (hunk 3 above)
Comment (mdhaber) [reply scipy/stats/_stats_py.py]: Committed!

File: scipy/stats/_stats_py.py (hunk 2 above)
Comment (HDembinski) [reply scipy/stats/_stats_py.py:7241]: I understand that the text not become too long, but in an ideal world, we should fully specify when exactly the test is applicable. We cannot make strong statements about finite samples, but we can fully specify when the test is applicable in the asymptotic regime (namely when the conditions are met under which it was originally derived). I think that would take little space.

Firstly, it is common in practice people apply statistical tests and methods to situations where they are not valid, because - understandably - one has to study the theory and read a lot of text books to really understand everything.

Secondly, we should be precise. This is a scientific library about applied math. It should be all about correctness. People often perceive statistics as a messy branch of math, but it is not messy at all. It is as well founded in proofs as any other branch of math.

File: scipy/stats/_stats_py.py (hunk 3 above)
Comment (HDembinski) [reply scipy/stats/_stats_py.py]: There is only one "multinomial distribution". The term is well-defined. If it is not what https://en.wikipedia.org/wiki/Multinomial_distribution describes, then it is not a multinomial distribution.

File: scipy/stats/_stats_py.py (hunk 4 above)
Comment (HDembinski) [reply scipy/stats/_stats_py.py:7229]: I like "not well-calibrated" although that implies that the reader knows what "calibration" means in this context, so I am also fine with the more lax term "invalid" or "not accurate".

**Comment by mdhaber** [comment]
Looks like we have three approvals of the essential points, so I'll go ahead and merge this one. Regarding the finer points of the rest of the documentation, feel free to open a PR. Thanks, all!
